File: enrich_raions.py
import json
from shapely.geometry import shape, Point, MultiPolygon, Polygon

def enrich_geojson(oblasts_path, raions_path, output_path):
    with open(oblasts_path, 'r', encoding='utf-8') as f:
        oblasts = json.load(f)
    
    with open(raions_path, 'r', encoding='utf-8') as f:
        raions = json.load(f)

    # Convert oblast features to shapely geometries
    oblast_geoms = []
    for feature in oblasts['features']:
        geom = shape(feature['geometry'])
        name = feature['properties']['NAME_1']
        oblast_geoms.append((name, geom))

    found_count = 0
    total_count = len(raions['features'])

    for raion in raions['features']:
        raion_geom = shape(raion['geometry'])
        raion_centroid = raion_geom.centroid
        
        assigned_oblast = "Unknown"
        max_intersection = 0
        
        # Determine parent oblast by max intersection area (robust for overlapping boundaries)
        for name, geom in oblast_geoms:
            if geom.contains(raion_centroid) or geom.intersects(raion_geom):
                intersection_area = geom.intersection(raion_geom).area
                if intersection_area > max_intersection:
                    max_intersection = intersection_area
                    assigned_oblast = name
        
        raion['properties']['parent_oblast'] = assigned_oblast
        if assigned_oblast != "Unknown":
            found_count += 1

    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(raions, f, ensure_ascii=False, indent=2)

    print(f"Enrichment complete. Mapped {found_count}/{total_count} raions.")

if __name__ == "__main__":
    enrich_geojson(
        '/home/ihor/Desktop/map/mapPractice/react-app/src/components/UKR_adm1.json',
        '/home/ihor/Desktop/map/mapPractice/react-app/src/components/UKR_adm2.json',
        '/home/ihor/Desktop/map/mapPractice/react-app/src/components/UKR_adm2_enriched.json'
    )

File: enrich_raions.js
import fs from 'fs';

function isPointInPolygon(point, polygon) {
    let x = point[0], y = point[1];
    let inside = false;
    for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
        let xi = polygon[i][0], yi = polygon[i][1];
        let xj = polygon[j][0], yj = polygon[j][1];
        let intersect = ((yi > y) !== (yj > y)) && (x < (xj - xi) * (y - yi) / (yj - yi) + xi);
        if (intersect) inside = !inside;
    }
    return inside;
}

function getCentroid(geometry) {
    let coords = [];
    if (geometry.type === 'Polygon') {
        coords = geometry.coordinates[0];
    } else if (geometry.type === 'MultiPolygon') {
        // Just use the first large polygon's coords for centroid
        coords = geometry.coordinates[0][0];
    }

    let x = 0, y = 0;
    for (let p of coords) {
        x += p[0];
        y += p[1];
    }
    return [x / coords.length, y / coords.length];
}

const oblasts = JSON.parse(fs.readFileSync('react-app/src/components/UKR_adm1.json', 'utf8'));
const raions = JSON.parse(fs.readFileSync('react-app/src/components/UKR_adm2.json', 'utf8'));

const oblastGeoms = oblasts.features.map(f => ({
    name: f.properties.NAME_1,
    geometry: f.geometry
}));

raions.features.forEach(raion => {
    const centroid = getCentroid(raion.geometry);
    let parent = "Unknown";

    for (const oblast of oblastGeoms) {
        let inside = false;
        if (oblast.geometry.type === 'Polygon') {
            inside = isPointInPolygon(centroid, oblast.geometry.coordinates[0]);
        } else if (oblast.geometry.type === 'MultiPolygon') {
            for (const poly of oblast.geometry.coordinates) {
                if (isPointInPolygon(centroid, poly[0])) {
                    inside = true;
                    break;
                }
            }
        }

        if (inside) {
            parent = oblast.name;
            break;
        }
    }
    raion.properties.parent_oblast = parent;
});

fs.writeFileSync('react-app/src/components/UKR_adm2.json', JSON.stringify(raions, null, 2));
console.log('Enrichment complete.');
